File: XUCore/additions/NSAttributedStringAdditions.m
// 
// NSAttributedStringAdditions.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-13 Fuel Collective, LLC. All rights reserved.
// 


#import "NSAttributedStringAdditions.h"


@implementation NSAttributedString (FCAdditions)

- (NSArray *)allAttachments{
	NSMutableArray *theAttachments = [NSMutableArray array];
	NSRange theStringRange = NSMakeRange(0, [self length]);
	if (theStringRange.length > 0)
	{
		NSUInteger N = 0;
		do
		{
			NSRange theEffectiveRange;
			NSDictionary *theAttributes = [self attributesAtIndex:N longestEffectiveRange:&theEffectiveRange inRange:theStringRange];
			NSTextAttachment *theAttachment = [theAttributes objectForKey:NSAttachmentAttributeName];
			if (theAttachment != NULL)
				[theAttachments addObject:theAttachment];
			N = theEffectiveRange.location + theEffectiveRange.length;
		}
		while (N < theStringRange.length);
	}
	return(theAttachments);
}

@end

File: XUCore/additions/NSAttributedStringAdditions.swift
//
//  NSAttributedStringAdditions.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/9/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation


public extension NSAttributedString {
	
	/// Returns all text attachments within self
	public var allAttachments: [NSTextAttachment] {
		var attachments: [NSTextAttachment] = [ ]
		
		let stringRange = NSMakeRange(0, self.length)
		if stringRange.length > 0 {
			var N = 0
			repeat {
				var theEffectiveRange: NSRange = NSMakeRange(0, 0)
				let theAttributes = self.attributes(at: N, longestEffectiveRange: &theEffectiveRange, in: stringRange)
				if let attachment = theAttributes[NSAttachmentAttributeName] as? NSTextAttachment {
					attachments.append(attachment)
				}
				
				N = theEffectiveRange.location + theEffectiveRange.length
			} while N < stringRange.length
		}
		
		return attachments
	}
	
}


